real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/localizers/KfLocalizer.java

```diff
@@ -1,43 +1,308 @@
+						double currentEncoderMeasurement = measurement
+								.getDistance();
+						double deltaDistance = currentEncoderMeasurement
+								- lastEncoderReading;
+						long currentTime = new Date().getTime();
+						long dt = currentTime - lastEncoderReadingTime;
+						if (dt > 1) { // to remove numeric instability
+							double bodySpeed = deltaDistance / (dt / 1000.0);
+							lastEncoderReadingTime = currentTime;
+							lastEncoderReading = currentEncoderMeasurement;
+
+							double[][] observationModel = {
+									{ 0, 0, 0, 0, 0, 0, 0 }, // x
+									{ 0, 0, 0, 0, 0, 0, 0 }, // y
+									{ 0, 0, 1, 0, 0, 0, 0 }, // x_dot_b
+									{ 0, 0, 0, 0, 0, 0, 0 }, // y_dot_b
+									{ 0, 0, 0, 0, 0, 0, 0 }, // th
+									{ 0, 0, 0, 0, 0, 0, 0 }, // th_dot
+									{ 0, 0, 0, 0, 0, 0, 0 } // Heading
+							};
+
+							double[][] meassurement = { { 0 }, { 0 },
+									{ bodySpeed }, { 0 }, { 0 }, { 0 }, { 0 } };
+							double[][] updateCovariance = {
+									{ 1, 0, 0, 0, 0, 0, 0 }, // x
+									{ 0, 1, 0, 0, 0, 0, 0 }, // y
+									{ 0, 0, 1, 0, 0, 0, 0 }, // x_dot
+									{ 0, 0, 0, 1, 0, 0, 0 }, // y_dot
+									{ 0, 0, 0, 0, 1, 0, 0 }, // th
+									{ 0, 0, 0, 0, 0, 1, 0 }, // th_dot
+									{ 0, 0, 0, 0, 0, 0, 1 } // heading
+							};
+							updateStep(new Matrix(observationModel),
+									new Matrix(meassurement), new Matrix(
+											updateCovariance));
+						}
+					}
+				});
+
+		new Subscriber("htGpsLoc", NodeChannel.STEERING.getMsgPath(),
+				new MessageListener() {
+
+					@Override
+					public synchronized void actionPerformed(String topicName,
+							Message m) {
+						SteeringMeasurement steerM = (SteeringMeasurement) m;
+						double[][] observationModel = {
+								{ 0, 0, 0, 0, 0, 0, 0 }, // x
+								{ 0, 0, 0, 0, 0, 0, 0 }, // y
+								{ 0, 0, 0, 0, 0, 0, 0 }, // x_dot_b
+								{ 0, 0, 0, 0, 0, 0, 0 }, // y_dot_b
+								{ 0, 0, 0, 0, 0, 0, 0 }, // th
+								{ 0, 0, 0, 0, 0, 0, 0 }, // th_dot
+								{ 0, 0, 0, 0, 0, 0, 1 } // Heading
+						};
+						double[][] meassurement = { { 0 }, { 0 }, { 0 }, { 0 },
+								{ 0 }, { 0 }, { steerM.getAngle() } };
+						double[][] updateCovariance = {
+								{ 1, 0, 0, 0, 0, 0, 0 }, // x
+								{ 0, 1, 0, 0, 0, 0, 0 }, // y
+								{ 0, 0, 1, 0, 0, 0, 0 }, // x_dot
+								{ 0, 0, 0, 1, 0, 0, 0 }, // y_dot
+								{ 0, 0, 0, 0, 1, 0, 0 }, // th
+								{ 0, 0, 0, 0, 0, 1, 0 }, // th_dot
+								{ 0, 0, 0, 0, 0, 0, 1 } // heading
+						};
+						updateStep(new Matrix(observationModel), new Matrix(
+								meassurement), new Matrix(updateCovariance));
+					}
+				});
+		resume();
+
+	}
+
+	private synchronized void updateStep(Matrix observationMatrix,
+			Matrix measurement, Matrix updateCovariance) {
+		predictStep();
+
+		// Matrix update = observationMatrix.times(state);
+		Matrix inovation = measurement.minus(observationMatrix.times(state));
+
+		for (int i = 4; i < 7; i++) {
+			if (inovation.get(i, 0) > 180) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2684047960/issues/source?bid=2938665&fileBranchId=3246677#l275)